### PR TITLE
MPI support for more algorithms.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Add MPI support for some algorithms required in the SNS powder reduction workflow. This does not affect the existing SNS MPI support and is only relevant when compiling with `-DMPI_EXPERIMENTAL=ON`.

**To test:**

Code review.
No release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/MatrixWorkspace.cpp
Comment (OwenArnold) [inline Framework/API/src/MatrixWorkspace.cpp]: Very minor point. If this is the kind of query that would be used repeatedly, would it be better to add something like `bool isDistributed() const` to `IndexInfo`?

**Review by OwenArnold — APPROVED** [review]
Needs approval from someone at the SNS before merging.

File: Framework/API/src/MatrixWorkspace.cpp
Comment (SimonHeybrock) [reply Framework/API/src/MatrixWorkspace.cpp]: For now just here. I do not see this being used a lot.

**Comment by SimonHeybrock** [comment]
@OwenArnold What SNS-specific changes are you referring to?

**Comment by OwenArnold** [comment]
@SimonHeybrock There's nothing specific that I can see, and system tests are passing, but since the SNS would be affected by anything overlooked that automated testing is not accounting for, someone like @peterfpeterson might want the opportunity to take a glance at this.

**Comment by martyngigg** [comment]
@SimonHeybrock The development documentation PR moved the MPI document to the `dev-docs` project so there are some conflicts here now.

**Comment by SimonHeybrock** [comment]
@martyngigg Thanks, fixed!

(it is always nice to see `git` magic at work: it managed rebasing automatically and applied all patches to the moved file correctly, without requiring intervention)

**Pull request merged**
